securedrop/journalist_app/admin.py

```diff
@@ -33,9 +35,30 @@ def manage_config():
             f = form.logo.data
             static_filepath = os.path.join(config.SECUREDROP_ROOT,
                                            "static/i/logo.png")
+            temp_static_filepath = os.path.join(config.SECUREDROP_ROOT,
+                                                "static/i/temp_logo.png")
+            with open(static_filepath) as img_file:
+                current_img = img_file.read()
             f.save(static_filepath)
-            flash(gettext("Image updated."), "logo-success")
-            return redirect(url_for("admin.manage_config"))
+
+            try:
+                with Image.open(static_filepath) as im:
+                    imcopy = im.copy()
+                    imcopy.thumbnail((500, 450), resample=3)
+                    imcopy.save(temp_static_filepath, "PNG")
+                if os.path.exists(static_filepath):
+                    os.rename(temp_static_filepath, static_filepath)
+                flash(gettext("Image updated."), "logo-success")
+
+            except IOError:
+                with open(static_filepath, 'w') as img_file:
```

Comment:
This `with` block is unnecessary becuse `os.rename` is effectively atomic, so I think it's acceptable to assume that we never end up with a broken `logo.png`.

Reply:
Since `f.save(static_filepath)` always writes logo.png, this `with` block replaces it with current image id it is broken. Without this restoration, only a message will be flashed, but the broken image will remain saved as logo.png

Reply:
Yes, which is why I'm suggesting that we have an endpoint that selects and uses something called `custom_logo.png` so we don't end up with a broken `logo.png`. If the app terminates while the write in that `with` block happens, we'll have a broken logo. You're more IO than necessary, and that could be a problem. Also, the advantage of having and endpoint that selects `custom_logo.png` is that if someone runs the Ansible playbooks, they won't overwrite the uploaded image.

Reply:
Before moving on to logo endpoint approach, I want to discuss my current approach and what purpose the two `with` blocks are serving.

Case 1-

User tries to upload logo for the first time
The image is a valid png file - A.png
The default SecureDrop logo is backed up
`f.save` writes A.png
In the `try` block, no exception occurs while working with image, and `os.rename()` executes
Thus the resized image is written at `static_filepath`

Case 2-

User takes a non-image file B.png (by renaming any other type of file to png) and tries to upload
The current logo is backed up by first `with` block
`f.save` writes B.png (which is a valid file, but not a valid image file)
Pillow raises `IOError` in the `try` block while trying to read the file, and `os.rename()` does not execute
In the `except` block, second `with` overwrites `static_filepath` with previously backed up image, which is not broken
Thus, the broken image is discarded, and a warning is shown

If the `with` blocks are removed, the non-image file will be saved. So I they are not unnecessary in the current approach.